### PR TITLE
spirv-fuzz: Refactor fuzzer, replayer and shrinker

### PR DESCRIPTION
**Pull request by afd**
In preparation for some upcoming work on the shrinker, this PR changes
the interfaces of Fuzzer, Replayer and Shrinker so that all data
relevant to each class is provided on construction, meaning that the
"Run" method can become a zero-argument method that returns a status,
transformed binary and sequence of applied transformations via a
struct.

This makes greater use of fields, so that -- especially in Fuzzer --
there is a lot less parameter passing.

### CODE REVIEW

**Comment by afd** [comment]
@Vasniktel If you have time to review this then I'd be keen to get feedback on the way I've used std::move.  I'm not yet super-confident on when it's best to use std::move.

File: source/fuzz/shrinker.cpp [hunk 1]
```diff
@@ -237,22 +242,30 @@ Shrinker::ShrinkerResultStatus Shrinker::Run(
...
   // step limit.
+  //
+  // Either way, the output from the shrinker is the best binary we saw, and the
+  // transformations that led to it.
   assert(attempt <= step_limit_);
   if (attempt == step_limit_) {
     std::stringstream strstream;
     strstream << "Shrinking did not complete; step limit " << step_limit_
               << " was reached.";
     consumer_(SPV_MSG_WARNING, nullptr, {}, strstream.str().c_str());
```
Comment (Vasniktel) [inline source/fuzz/shrinker.cpp:255]: According to https://en.cppreference.com/w/cpp/io/basic_stringstream/str, `str()` returns a temporary object. Thus, `str().c_str()` results in a dangling pointer.

I've seen this bug somewhere else in the project as well. Maybe it would be better to make a separate PR for fixing it.

File: source/fuzz/shrinker.cpp [hunk 2]
```diff
@@ -237,22 +242,30 @@ Shrinker::ShrinkerResultStatus Shrinker::Run(
...
-    return Shrinker::ShrinkerResultStatus::kStepLimitReached;
+    return {Shrinker::ShrinkerResultStatus::kStepLimitReached,
+            current_best_binary, current_best_transformations};
   }
-  return Shrinker::ShrinkerResultStatus::kComplete;
+  return {Shrinker::ShrinkerResultStatus::kComplete, current_best_binary,
+          current_best_transformations};
+}
```
Comment (Vasniktel) [inline source/fuzz/shrinker.cpp:261]: You could've probably used `std::move` here as well.

File: source/fuzz/fuzzer.h [hunk 3]
```diff
@@ -70,73 +75,64 @@ class Fuzzer {
...
-      const std::vector<fuzzerutil::ModuleSupplier>& donor_suppliers,
-      std::vector<uint32_t>* binary_out,
-      protobufs::TransformationSequence* transformation_sequence_out);
+  // Transforms |binary_in_| by running a number of randomized fuzzer passes.
+  // Initial facts about the input binary and the context in which it will
+  // execute are provided via |initial_facts_|.  A source of donor modules to be
+  // used by transformations is provided via |donor_suppliers_|.  On success,
+  // returns the transformed binary and the sequence of transformations that
+  // were applied.
+  FuzzerResult Run();
```
Comment (Vasniktel) [inline source/fuzz/fuzzer.h:85]: Maybe it's better to explicitly mention here that `FuzzerResult::trasformed_binary` and `FuzzerResult::applied_transformations` are empty if `FuzzerResult::status` is not `kComplete`. @afd, what do you think?

If you agree, don't forget to adjust your comments in shrinker and replayer as well.

**Review by Vasniktel — COMMENTED** [review]
Looks good. Your uses of `std::move` seem OK to me.

File: source/fuzz/shrinker.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/shrinker.cpp:255]: I believe this is OK because the lifetime of a temporary object is dictated by the "full-expression" in which it is nested.

Here, the full-expression is:

```
consumer_(SPV_MSG_WARNING, nullptr, {}, strstream.str().c_str())
```

I believe there would only be UB if we did something like this instead:

```
auto temp = strstr.str().c_str();
consumer_(SPV_MSG_WARNING, nullptr, {}, temp)
```

File: source/fuzz/shrinker.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/shrinker.cpp:261]: Done.

File: source/fuzz/fuzzer.h (hunk 3 above)
Comment (afd) [reply source/fuzz/fuzzer.h:85]: Done.

File: source/fuzz/shrinker.cpp (hunk 1 above)
Comment (Vasniktel) [reply source/fuzz/shrinker.cpp:255]: Yes, you are right! I didn't know that. In this case, I'm OK with the current approach.

**Review by Vasniktel — APPROVED** [review]

File: source/fuzz/shrinker.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/shrinker.cpp:255]: Thanks.
